### ISSUE TITLE
nRF5340 flashing issue: Failed to prepare for programming. Could not preserve target memory.

### ISSUE THREAD

**Issue by vpmdt** [square/pylink#177]
Hi,
struggling with flashing nRF5340. 
If I flash it manually with J-Flash Lite, first NET then APP core, it works fine.
I need to flash both cores with no reset between them,  NET and APP respectively, because of the readback protection.

```
jlk = pylink.JLink()
jlk.open(serial_no=int(jlink_sn))
jlk.set_tif(pylink.enums.JLinkInterfaces.SWD)
....
....
    jlk.connect(chip_name='nrf5340_xxaa_net', verbose=True)
    jlk.flash(binstr_obj1, min_addr1)

    jlk.connect(chip_name='nrf5340_xxaa_app', verbose=True)
    jlk.flash(binstr_obj2, min_addr2)
```
first flash is succesfully done, but after second flash I get error:
Failed to prepare for programming.
Could not preserve target memory.

Any ideas? Thanks!!

**Comment by hkpeprah** [comment]
Do you have any more logs? Is it possible that the chip is not being properly held in reset?

**Comment by vpmdt** [comment]
I'm trying more things to get it working, so here is the only succesful setup for flashing:

1. First the manual flash of merged hex (net and app core) is done via nRF Programmer
2.  Then I can run the script which is
`jlk = pylink.JLink()
jlk.open(serial_no=int(jlink_sn))
jlk.set_tif(pylink.enums.JLinkInterfaces.SWD)

# SWDSelect
jlk.coresight_configure()

# SWDWriteDP 1 0x50000000
jlk.coresight_write(reg=1, data=0x50000000, ap=False)

# SWDWriteDP 2 0x020000F0
jlk.coresight_write(reg=2, data=0x020000F0, ap=False)

# SWDReadAP 3
jlk.coresight_read(reg=3, ap=True)
time.sleep(0.5)

# SWDReadAP 3
jlk.coresight_read(reg=3, ap=True)
time.sleep(0.5)

# SWDWriteDP 2 0x02000000
jlk.coresight_write(reg=2, data=0x02000000, ap=False)

# SWDWriteAP 1 1
jlk.coresight_write(reg=1, data=1, ap=True)

# SWDReadAP 2
jlk.coresight_read(reg=2, ap=True)
time.sleep(0.5)

# SWDReadAP 2
jlk.coresight_read(reg=2, ap=True)
time.sleep(0.5)

# SWDReadAP 2
jlk.coresight_read(reg=2, ap=True)
time.sleep(0.5)

# SWDWriteDP 2 0x03000000
jlk.coresight_write(reg=2, data=0x03000000, ap=False)

# SWDWriteAP 1 1
jlk.coresight_write(reg=1, data=1, ap=True)

# SWDReadAP 2
jlk.coresight_read(reg=2, ap=True)
time.sleep(0.5)

# SWDReadAP 2
jlk.coresight_read(reg=2, ap=True)
time.sleep(0.5)

# SWDReadAP 2
jlk.coresight_read(reg=2, ap=True)
time.sleep(0.5)


jlk.connect(chip_name='nrf5340_xxaa_net', verbose=True)
jlk.flash(binstr_obj1, min_addr1)

jlk.connect(chip_name='nrf5340_xxaa_app', verbose=True)
jlk.flash(binstr_obj2, min_addr2)
`

If the chip is fully erased it fails as described in my first comment - it fails when the second flash is executed (when the app core is being flashed). Let me know if you need other information. Thank you!

**Comment by enjeru15** [comment]
Hi, has this been resolved already?

**Comment by ahmedyahia3393** [comment]
Hi
While flashing a hex file, I have the same error but using nrf52840 and Segger j-link. you can see the attached photo as a reference. I tried resetting the chip and erasing it beforehand, but nothing changed. 
Thanks. 
<img width="665" alt="flashing error" src="https://github.com/square/pylink/assets/38445299/02827933-5527-4c40-aa99-e1f87010b51b">

**Comment by hkpeprah** [comment]
Could you try a different reset strategy? I don't have an NRF5340 to play around with, unfortunately.

**Comment by juhhov** [comment]
Pylink does not allow connecting to second core due to this https://github.com/square/pylink/blob/0b3e8ca8768725b27b1fd58df71f7110084ee6fb/pylink/jlink.py#L1156C1-L1159C52.
Once first core is connected to the attempt to connect to second core is refused silently.

**Comment by hkpeprah** [comment]
> Pylink does not allow connecting to second core due to this https://github.com/square/pylink/blob/0b3e8ca8768725b27b1fd58df71f7110084ee6fb/pylink/jlink.py#L1156C1-L1159C52. Once first core is connected to the attempt to connect to second core is refused silently.

Interesting. Do you have example code? Is the second `Device =` call supposed to fail? Maybe we're missing a disconnect, or need to instantiate another instance and tear down the existing one.

**Comment by juhhov** [comment]
Yes, pylink is missing disconnect API (or something similar) to disconnect target connection. Adding such API might be the least effort?

If exec_command and straight DLL access is used to bypass pylink API's at least the flashing of second core seems to succeed.

```
jlink.exec_command('Device = NRF5340_XXAA_NET')
jlink._dll.JLINKARM_Connect()
```

**Comment by hkpeprah** [comment]
I will take a look; this is something I over-looked as I haven't worked on multi-CPU ARM chips.

**Comment by juhhov** [comment]
Thanks. The workaround works for us for now. I can definitely test your PR.
